src/libstd/sys/unix/ext/net/ancillary.rs

```diff
@@ -0,0 +1,615 @@
+        target_os = "ios",
+        target_os = "freebsd",
+        target_os = "dragonfly",
+        target_os = "openbsd",
+        target_os = "netbsd",
+        target_os = "linux",
+        target_os = "android",
+        target_os = "emscripten",
+        target_os = "fuchsia",
+        target_env = "uclibc",
+    ))]
+    fn as_rights(data: &'a [u8]) -> Self {
+        let ancillary_data_iter = AncillaryDataIter::new(data);
+        let scm_rights = ScmRights(ancillary_data_iter);
+        AncillaryData::ScmRights(scm_rights)
+    }
+
+    #[cfg(any(
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "freebsd",
+        target_os = "dragonfly",
+        target_os = "macos",
+        target_os = "ios",
+        target_os = "linux",
+        target_os = "android",
+        target_os = "emscripten",
+        target_os = "fuchsia",
+        target_env = "uclibc",
+    ))]
+    fn as_credentials(data: &'a [u8]) -> Self {
+        let ancillary_data_iter = AncillaryDataIter::new(data);
+        let scm_credentials = ScmCredentials(ancillary_data_iter);
+        AncillaryData::ScmCredentials(scm_credentials)
+    }
+}
+
+#[cfg(any(
+    target_os = "haiku",
+    target_os = "solaris",
+    target_os = "illumos",
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "freebsd",
+    target_os = "dragonfly",
+    target_os = "openbsd",
+    target_os = "netbsd",
+    target_os = "linux",
+    target_os = "android",
+    target_os = "emscripten",
+    target_os = "fuchsia",
+    target_env = "uclibc",
+))]
+#[unstable(feature = "unix_socket_ancillary_data", issue = "none")]
+impl<'a> AncillaryData<'a> {
+    fn from(cmsg: &'a libc::cmsghdr) -> Self {
+        unsafe {
+            let cmsg_len_zero = libc::CMSG_LEN(0) as usize;
+            let data_len = (*cmsg).cmsg_len - cmsg_len_zero;
+            let data = libc::CMSG_DATA(cmsg).cast();
+            let data = from_raw_parts(data, data_len);
+
+            if (*cmsg).cmsg_level == libc::SOL_SOCKET {
+                match (*cmsg).cmsg_type {
+                    libc::SCM_RIGHTS => AncillaryData::as_rights(data),
+                    #[cfg(any(
+                        target_os = "linux",
+                        target_os = "android",
+                        target_os = "emscripten",
+                        target_os = "fuchsia",
+                        target_env = "uclibc",
+                    ))]
+                    libc::SCM_CREDENTIALS => AncillaryData::as_credentials(data),
+                    #[cfg(any(
+                        target_os = "netbsd",
+                        target_os = "openbsd",
+                        target_os = "freebsd",
+                        target_os = "dragonfly",
+                        target_os = "macos",
+                        target_os = "ios",
+                    ))]
+                    libc::SCM_CREDS => AncillaryData::as_credentials(data),
+                    _ => panic!("Unknown cmsg type"),
```

Comment:
This shouldn't panic on an unknown cmsg type; it should return an error to its caller, and the caller should pass that error through to the user. That allows a program to recover from receiving an unexpected ancillary data type (and continue on to receive the next bit of ancillary data).